File: probs/14221.py
import sys
input = sys.stdin.readline

input = open('input.txt', 'r').readline

from heapq import heappush, heappop

INF = 10**12

def dijkstra(starts):
    dist = [INF]*(N+1)
    pq = []
    for s in starts:
        dist[s] = 0
        heappush(pq, (0, s))

    while pq:
        _, s = heappop(pq)
        for t, d in edges[s].items():
            if dist[t] > dist[s] + d:
                dist[t] = dist[s] + d
                heappush(pq, (dist[t], t))
    return dist

N, M = map(int, input().split())

edges = [{} for _ in range(N+1)]
for _ in range(M):
    u, v, c = map(int, input().split())
    if v in edges[u]:
        edges[u][v] = min(edges[u][v], c)
    else:
        edges[u][v] = c
    edges[v][u] = edges[u][v]

input()

cands = list(map(int, input().split()))
stores = list(map(int, input().split()))

m = INF
for i, d in enumerate(dijkstra(stores)):
    if i in cands and m > d:
        m = d
        ans = i
print(ans)

File: probs/14221.cpp
#include <bits/stdc++.h>

using namespace std;
using pi = pair<int, int>;
using ti = tuple<int, int, int>;

const int MAX = 5000,
          INF = 1e9;

int dist[MAX+1], H[MAX], C[MAX];
map<int, int> G[MAX+1];

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    freopen("input.txt", "r", stdin);

    int V, E; cin >> V >> E;
    while (E--) {
        int u, v, w; cin >> u >> v >> w;
        if (G[u].find(v) == G[u].end()) G[u][v] = G[v][u] = w;
        else G[u][v] = G[v][u] = min(G[u][v], w);
    }

    fill(dist, dist+V+1, INF);
    priority_queue<pi, vector<pi>, greater<pi>> pq;

    int p, q; cin >> p >> q;
    for (int i = 0; i < p; i++) cin >> H[i];
    sort(H, H+p);
    for (int i = 0; i < q; i++) {
        cin >> C[i];
        dist[C[i]] = 0;
        pq.emplace(0, C[i]);
    }

    while (!pq.empty()) {
        auto [d, s] = pq.top(); pq.pop();

        if (dist[s] < d) continue;

        for (auto [t, w] : G[s]) {
            if (dist[t] > d+w) {
                dist[t] = d+w;
                pq.emplace(d+w, t);
            }
        }
    }

    int ans, d = INF;
    for (int i = 0; i < p; i++) {
        if (dist[H[i]] < d) {
            d = dist[H[i]];
            ans = H[i];
        }
    }

    cout << ans << '\n';

    return 0;
}
